### PR TITLE
Rebuild Athena Partitions

### LINKED ISSUE #424: [bug] Do not allow '_' (underscores) in prefix
## Background

If prefix contains an underscore, terraform will attempt to create and S3 bucket with an underscore when it builds the infrastructure.  This will result in an invalid bucket name error due to bucket name restrictions.

## Desired Change

upon prefix creation, validate allowed characters for bucket names.

## Logs

```aws_cloudtrail.streamalert: InvalidS3BucketNameException: Bucket name should not contain '_'```

### PR DESCRIPTION
**Pull request by jacknagz**
to: @ryandeivert 
cc: @airbnb/streamalert-maintainers @securityclippy 
size: medium
resolves #424

## Background

There are a lot of disparate changes in this PR related to the CLI.  But it mainly is for adding capabilities for Athena to drop all tables (when destroying clusters), and rebuilding table partitions (when a schema is updated).

## Changes

* Int => BigInt for Hive column defaults
* Modify the GHE schema for elapsed to be a string, there were parsing issues with Athena
* Add a continue prompt when destroying, this does some damage if you accidentally type this command
* Write helper functions for Athena when interacting with output and rebuilding partitions
* Add support for rebuilding partitions
* Add support for dropping all tables
* Other small cleanup

## Testing

All tested in a side AWS account

### CODE REVIEW

File: stream_alert_cli/config.py [hunk 1]
```diff
@@ -113,6 +113,10 @@ def set_prefix(self, prefix):
             LOGGER_CLI.error('Invalid prefix type, must be string')
             return
 
+        if prefix.startswith('_'):
```
Comment (ryandeivert) [inline stream_alert_cli/config.py]: I think the prefix can't contain an underscore at all, not just start with it.

File: stream_alert_cli/runner.py [hunk 2]
```diff
@@ -41,8 +43,7 @@ def cli_runner(options):
             Contains the following keys for lambda commands:
                 (command, subcommand, env, func, source)
     """
-    cli_load_message = ('Issues? Report here: '
-                        'https://github.com/airbnb/streamalert/issues')
+    cli_load_message = ('Issues? Report here: ' 'https://github.com/airbnb/streamalert/issues')
```
Comment (ryandeivert) [inline stream_alert_cli/runner.py]: I think yapf did this, but the parenths are no longer required, along with the quotes in the middle

File: stream_alert_cli/terraform/handler.py [hunk 3]
```diff
@@ -110,6 +110,9 @@ def terraform_handler(options, config):
         terraform_clean(config)
 
     elif options.subcommand == 'destroy':
+        if not continue_prompt(message='Are you sure you want to destroy?'):
```
Comment (ryandeivert) [inline stream_alert_cli/terraform/handler.py]: ++ nice

**Comment by ryandeivert** [comment]
For future reference: we should consider adding a continue prompt for the `init` command to, maybe warning users it should only be done once at the very beginning, since that seems to be one that is easily ran without knowing the consequence

File: stream_alert_cli/config.py (hunk 1 above)
Comment (securityclippy) [reply stream_alert_cli/config.py]: This is correct.  an underscore in the middle of the prefix will also result in an issue in resource creation.  see https://github.com/airbnb/streamalert/issues/424.

Comment (jacknagz) [reply stream_alert_cli/config.py]: Ok! I'll make that change now

File: stream_alert_cli/runner.py (hunk 2 above)
Comment (jacknagz) [reply stream_alert_cli/runner.py]: Yep, it was `yapf`, not me I promise

**Comment by jacknagz** [comment]
Testing:

```
$ python manage.py  athena rebuild-partitions --table_name cloudwatch_events --bucket <prefix>.streamalert.data --type data
StreamAlertCLI [INFO]: Issues? Report here: https://github.com/airbnb/streamalert/issues
StreamAlertCLI [INFO]: Dropping table cloudwatch_events
StreamAlertCLI [INFO]: Dropped table cloudwatch_events
StreamAlertAthena [INFO]: The streamalert table 'cloudwatch_events' does not exist.
StreamAlertAthena [INFO]: For help with creating tables: $ python manage.py athena create-table --help
StreamAlertCLI [INFO]: The cloudwatch_events table was successfully created!
StreamAlertCLI [INFO]: Creating 31 new partitions for cloudwatch_events
StreamAlertCLI [INFO]: Successfully rebuilt partitions for cloudwatch_events
StreamAlertCLI [INFO]: Completed
```

**Comment by jacknagz** [comment]
Hey @ryandeivert @securityclippy just pushed an update, PTAL

File: stream_alert_cli/athena/handler.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    """Rebuild an Athena table's partitions
+
+    Steps:
+      - Get the list of current partitions
+      - Destroy existing table
+      - Re-create tables
+      - Re-create partitions
+
+    Args:
+        athena_client (boto3.client): Instantiated CLI AthenaClient
```
Comment (ryandeivert) [inline stream_alert_cli/athena/handler.py]: It looks like there are some args missing from the docstring here

File: stream_alert/athena_partition_refresh/helpers.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+
+def partition_statement(partitions, bucket, table_name):
+    """Generate ALTER TABLE commands from existing partitions.
+
+    Args:
+        partitions (set): The unique set of partitions gathered from Athena
+        bucket (str): The bucket name
+        table_name (str): Theh name of the Athena table
```
Comment (ryandeivert) [inline stream_alert/athena_partition_refresh/helpers.py]: `Theh` --> `The`

File: stream_alert_cli/athena/handler.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+            else:
+                athena_schema[key_name] = SCHEMA_TYPE_MAPPING[key_type]
+
+
+def create_table(athena_client, options, config):
+    """Create a 'streamalert' Athena table
+
+    Args:
+        athena_client (boto3.client): Instantiated CLI AthenaClient
+        options (namedtuple): The parsed args passed from the CLI
```
Comment (ryandeivert) [inline stream_alert_cli/athena/handler.py]: Missing config arg in list

File: stream_alert_cli/athena/handler.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+            LOGGER_CLI.error('Table name %s missing from configuration or '
+                             'is not enabled.', options.table_name)
+            return
+
+        if athena_client.check_table_exists(options.table_name):
+            LOGGER_CLI.info('The \'%s\' table already exists.', options.table_name)
+            return
+
+        log_info = config['logs'][options.table_name.replace('_', ':', 1)]
+        schema = dict(log_info['schema'])
```
Comment (ryandeivert) [inline stream_alert_cli/athena/handler.py]: isn't this already a dict? why is the casting necessary here?

File: stream_alert_cli/runner.py [hunk 8]
```diff
@@ -41,8 +38,7 @@ def cli_runner(options):
             Contains the following keys for lambda commands:
                 (command, subcommand, env, func, source)
     """
-    cli_load_message = ('Issues? Report here: '
-                        'https://github.com/airbnb/streamalert/issues')
+    cli_load_message = ('Issues? Report here: https://github.com/airbnb/streamalert/issues')
```
Comment (ryandeivert) [inline stream_alert_cli/runner.py]: doesn't pylint complain about unnecessary parenthesis here?

File: stream_alert_cli/config.py [hunk 9]
```diff
@@ -113,6 +113,10 @@ def set_prefix(self, prefix):
             LOGGER_CLI.error('Invalid prefix type, must be string')
             return
 
+        if '_' in prefix:
+            LOGGER_CLI.error('Prefix cannot start with underscore')
```
Comment (ryandeivert) [inline stream_alert_cli/config.py]: Please update the error message here

File: stream_alert_cli/runner.py (hunk 8 above)
Comment (jacknagz) [reply stream_alert_cli/runner.py]: apparently not, i'll fix though

File: stream_alert_cli/athena/handler.py (hunk 7 above)
Comment (jacknagz) [reply stream_alert_cli/athena/handler.py]: OrderedDict to dict

**Pull request merged**